### PR TITLE
MQTT fixes

### PR DESCRIPTION
**Pull request by davidgraeff**
* Fix MQTT broker network discovery
* Improve documentation
* Fix nullpointer in embedded broker (by introducing a reset for the metrics class)
* Add missing Null annotations
* Remove unused MQTT will and testament builder from the embedded broker package
* MQTTBrokerConnection: subscribe/unsubscibe future: Fail exceptionally instead of with a boolean value of false if the aforementioned actions fail. 

Signed-off-by: David Gräff <david.graeff@web.de>

### CODE REVIEW

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/ChannelGroupUID.java [hunk 1]
```diff
@@ -33,6 +33,12 @@
         super();
     }
 
+    /**
+     * Parses a {@link ChannelGroupUID} for a given string. The UID must be in the format
+     * 'bindingId:segment:segment:...'.
+     *
+     * @param channelGroupUid uid in form a string (must not be null)
+     */
```
Comment (davidgraeff) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/ChannelGroupUID.java]: I couldn't resist on commenting this constructor.

**Comment by htreu** [comment]
as Travis woke up, here is what he's got to say:

```
[ERROR] Failed to execute goal org.openhab.tools.sat:sat-plugin:0.5.0:checkstyle (sat-all) on project org.eclipse.smarthome.binding.mqtt: Unable to execute mojo: An error has occurred in Checkstyle report generation. Failed during checkstyle configuration: Exception was thrown while processing /home/travis/build/eclipse/smarthome/extensions/binding/org.eclipse.smarthome.binding.mqtt/src/main/java/org/eclipse/smarthome/binding/mqtt/internal/discovery/NetworkDiscoveryService.java: MismatchedTokenException occurred during the analysis of file /home/travis/build/eclipse/smarthome/extensions/binding/org.eclipse.smarthome.binding.mqtt/src/main/java/org/eclipse/smarthome/binding/mqtt/internal/discovery/NetworkDiscoveryService.java. expecting EOF, found '}' -> [Help 1]
```

**Comment by davidgraeff** [comment]
The PR #6327 should be applied first. It removes access to the ChannelDefinition constructor, which this PR uses and that need to be reworked.

Regarding checkstyle: It apparently can't handle this syntax: `protected MqttBrokerConnection testConnections @Nullable [];`

File: extensions/binding/org.eclipse.smarthome.binding.mqtt/src/main/java/org/eclipse/smarthome/binding/mqtt/handler/SystemBrokerHandler.java [hunk 2]
```diff
@@ -93,20 +95,21 @@ public void connectionStateChanged(MqttConnectionState state, @Nullable Throwabl
      * is no connection established yet.
      */
     @Override
-    public void brokerAdded(String connectionName, MqttBrokerConnection broker) {
-        if (!connectionName.equals(brokerID) || connection == broker) {
+    public void brokerAdded(String connectionName, MqttBrokerConnection addedConnectipn) {
```
Comment (htreu) [inline extensions/binding/org.eclipse.smarthome.binding.mqtt/src/main/java/org/eclipse/smarthome/binding/mqtt/handler/SystemBrokerHandler.java]: typo `addedConnectipn`

File: extensions/io/org.eclipse.smarthome.io.mqttembeddedbroker/src/main/java/org/eclipse/smarthome/io/mqttembeddedbroker/internal/MqttEmbeddedBrokerUserAuthenticator.java [hunk 3]
```diff
@@ -32,7 +36,7 @@ public MqttEmbeddedBrokerUserAuthenticator(String username, byte[] password) {
     }
 
     @Override
-    public boolean checkValid(String clientId, String username, byte[] password) {
+    public boolean checkValid(@Nullable String clientId, @Nullable String username, byte @Nullable [] password) {
         return this.username.equals(username) && Arrays.equals(this.password, password);
     }
 }
```
Comment (htreu) [inline extensions/io/org.eclipse.smarthome.io.mqttembeddedbroker/src/main/java/org/eclipse/smarthome/io/mqttembeddedbroker/internal/MqttEmbeddedBrokerUserAuthenticator.java]: missing new line

File: bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttBrokerConnectionConfig.java [hunk 4]
```diff
@@ -47,20 +48,20 @@
...
             StringBuffer b = new StringBuffer();
             if (host != null) {
                 b.append(host);
             }
+            final Integer port = this.port;
             if (port != null) {
                 b.append(":");
                 b.append(port.toString());
             }
-            if (secure != null) {
```
[the remark this answers is not in the document]
Comment (davidgraeff) [reply bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttBrokerConnectionConfig.java]: correct is of course :
```
if (secure) {
 b.append(":s");
}
```

**Pull request merged**